go/ct/gen/block_context_test.go

```diff
@@ -54,56 +56,476 @@ func TestBlockContextGen_Generate(t *testing.T) {
+		t.Run(name, func(t *testing.T) {
+			generator := NewBlockContextGenerator()
+			assignment := Assignment{}
+			test.setup(generator, assignment)
+			res, err := generator.Generate(assignment, randomSource)
+			if err != nil {
+				t.Fatalf("Error generating result for constraints %v and assignment %v: %v", generator, assignment, err)
+			}
+			test.check(t, res, assignment)
+		})
+	}
+}
+
+func TestBlockContextGenerator_SignalsUnsatisfiableForUnsatisfiableConstraints(t *testing.T) {
+	// TODO: add support for pre-assigned values
+	tests := map[string]func(*BlockContextGenerator, Assignment){
+		"conflicting-revisions": func(b *BlockContextGenerator, _ Assignment) {
+			b.SetRevision(common.R07_Istanbul)
+			b.SetRevision(common.R09_Berlin)
+		},
+		"conflicting-ranges-in-first": func(b *BlockContextGenerator, _ Assignment) {
+			b.RestrictVariableToOneOfTheLast256Blocks("a")
+			b.RestrictVariableToNoneOfTheLast256Blocks("a")
+		},
+		"conflicting-ranges-out-first": func(b *BlockContextGenerator, _ Assignment) {
+			b.RestrictVariableToNoneOfTheLast256Blocks("a")
+			b.RestrictVariableToOneOfTheLast256Blocks("a")
+		},
+		"conflicting-fixed-values": func(b *BlockContextGenerator, _ Assignment) {
+			b.SetBlockNumberOffsetValue("a", 44)
+			b.SetBlockNumberOffsetValue("a", 45)
+		},
+		"conflicting-fixed-values-with-out-of-range": func(b *BlockContextGenerator, _ Assignment) {
+			b.SetBlockNumberOffsetValue("a", 44)
+			b.RestrictVariableToNoneOfTheLast256Blocks("a")
+		},
+		"conflicting-fixed-values-with-in-range": func(b *BlockContextGenerator, _ Assignment) {
+			b.SetBlockNumberOffsetValue("a", 400)
+			b.RestrictVariableToOneOfTheLast256Blocks("a")
+		},
+		"block-number-overflow": func(b *BlockContextGenerator, a Assignment) {
+			b.SetBlockNumberOffsetValue("a", 400)
+			a["a"] = common.NewU256(1, 500) // 2^64+400
+		},
+		"block-number-underflow": func(b *BlockContextGenerator, a Assignment) {
+			b.SetRevision(common.R07_Istanbul)
+			b.SetBlockNumberOffsetValue("a", 1100)
+		},
+		"conflicting-revisions-with-in-range-and-predefined-assignment": func(b *BlockContextGenerator, a Assignment) {
+			b.SetRevision(common.R07_Istanbul)
+			b.RestrictVariableToOneOfTheLast256Blocks("a")
+			a["a"] = common.NewU256(8000)
+		},
+		"conflicting-revisions-with-in-range-and-predefined-assignment-bigger-than-uint64": func(b *BlockContextGenerator, a Assignment) {
+			b.SetRevision(common.R07_Istanbul)
+			b.RestrictVariableToOneOfTheLast256Blocks("a")
+			a["a"] = common.NewU256(1, 1)
+		},
+		"conflicting-revisions-with-in-range-and-predefined-assignment-max-uint64": func(b *BlockContextGenerator, a Assignment) {
+			b.SetRevision(common.R07_Istanbul)
+			b.RestrictVariableToOneOfTheLast256Blocks("a")
+			a["a"] = common.NewU256(0xffffffffffffffff)
+		},
+		"coflicting-assigned-value-and-fixed-offset": func(b *BlockContextGenerator, a Assignment) {
+			b.SetBlockNumberOffsetValue("a", 440)
+			b.SetBlockNumberOffsetValue("a", 40)
+		},
+	}
```

Comment:
This test should be result in an unsatisfiable error, but it does not:
```
                "no-valid-block-range": func(b *BlockContextGenerator, a Assignment) {
			b.blockNumberSolver = NewRangeSolver[uint64](0, 0)
			b.RestrictVariableToOneOfTheLast256Blocks("a")
		},
```

This test case forces the resulting block number to be `0`, which is a valid block number if the constraint would be to be part of the Istanbul revision (range 0 to 1000). The random number generator picking the revision in this range may chose to pick 0.